easybuild/easyblocks/l/llvm.py

```diff
@@ -94,31 +305,611 @@ def configure_step(self):
             except KeyError:
                 raise EasyBuildError("No default build targets defined for CPU architecture %s.", arch)
 
-        unknown_targets = [target for target in build_targets if target not in CLANG_TARGETS]
+        unknown_targets = set(build_targets) - set(ALL_TARGETS)
 
         if unknown_targets:
             raise EasyBuildError("Some of the chosen build targets (%s) are not in %s.",
-                                 ', '.join(unknown_targets), ', '.join(CLANG_TARGETS))
+                                 ', '.join(unknown_targets), ', '.join(ALL_TARGETS))
+        exp_targets = set(build_targets) & set(LLVM_EXPERIMENTAL_TARGETS)
+        if exp_targets:
+            self.log.warning("Experimental targets %s are being used.", ', '.join(exp_targets))
 
-        self.cfg.update('configopts', '-DLLVM_TARGETS_TO_BUILD="%s"' % ';'.join(build_targets))
+        self.build_targets = build_targets or []
 
-        if LooseVersion(self.version) >= LooseVersion('15.0'):
-            # make sure that CMake modules are available in build directory,
-            # by moving the extracted folder to the expected location
-            cmake_modules_path = os.path.join(self.builddir, 'cmake-%s.src' % self.version)
-            if os.path.exists(cmake_modules_path):
-                move_file(cmake_modules_path, os.path.join(self.builddir, 'cmake'))
+        self.nvptx_cond = 'NVPTX' in self.build_targets
+        self.amd_cond = 'AMDGPU' in self.build_targets
+        self.all_cond = 'all' in self.build_targets
+        self.cuda_cc = []
+        if self.nvptx_cond or self.all_cond:
+            # list of CUDA compute capabilities to use can be specifed in two ways (where (2) overrules (1)):
+            # (1) in the easyconfig file, via the custom cuda_compute_capabilities;
+            # (2) in the EasyBuild configuration, via --cuda-compute-capabilities configuration option;
+            ec_cuda_cc = self.cfg['cuda_compute_capabilities']
+            cfg_cuda_cc = build_option('cuda_compute_capabilities')
+            cuda_cc = cfg_cuda_cc or ec_cuda_cc or []
+            if not cuda_cc and self.nvptx_cond:
+                raise EasyBuildError(
+                    "Can't build Clang with CUDA support without specifying 'cuda-compute-capabilities'"
+                    )
```

Comment:
Please correct me if I'm wrong:

The current approach would just take the host architecture and not NVPTX / AMDGPU if the user doesn't explicitly set anything in their EasyConfig, right? Would it make sense to use the approach from the [Clang EasyBlock](https://github.com/easybuilders/easybuild-easyblocks/blob/21ad3f57a7c40078a5402bf0cdb4be710ae49a15/easybuild/easyblocks/c/clang.py#L180) or detect enabling them in another way (e.g. via `cuda_compute_capabilities`)?

```py
if build_targets is None:
   deps = [dep['name'].lower() for dep in self.cfg.dependencies()]
   arch = get_cpu_architecture()
   try:
       default_targets = DEFAULT_TARGETS_MAP[arch][:]
       # If CUDA is included as a dep, add NVPTX as a target
       # There are (old) toolchains with CUDA as part of the toolchain
       cuda_toolchain = hasattr(self.toolchain, 'COMPILER_CUDA_FAMILY')
       if 'cuda' in deps or cuda_toolchain:
           default_targets += ['NVPTX']
       # For AMDGPU support we need ROCR-Runtime and
       # ROCT-Thunk-Interface, however, since ROCT is a dependency of
       # ROCR we only check for the ROCR-Runtime here
       # https://openmp.llvm.org/SupportAndFAQ.html#q-how-to-build-an-openmp-amdgpu-offload-capable-compiler
       if 'rocr-runtime' in deps:
           default_targets += ['AMDGPU']
       self.cfg['build_targets'] = build_targets = default_targets
       self.log.debug("Using %s as default build targets for CPU/GPU architecture %s.", default_targets, arch)
   except KeyError:
       raise EasyBuildError("No default build targets defined for CPU architecture %s.", arch)
```

Reply:
Yes currently only the host architecture is used by default.

What the `Clang` EB does is just check if `cuda` or `rocr-runtime` is in the dependencies.
I think atleast for CUDA i had removed it as i was able to get offload going without adding CUDA to the deps (will have to double check as i did it a couple months ago and don't have access to the frontend where i tested this).

For using `cuda_compute_capabilities` i followed the same approach as the `Clang` EB of checking that it is defined when `NVPTX` is enabled as a target and set the proper variables (or fail if it is not defined).

I guess it would make sense to automatically enable NVPTX/AMDGPU if the respective deps are specified in the EC or if `cuda_compute_capabilities`/`amd_gfx_list` are defined

I am thinking the advantage of the automatic detection would be to be able to have NVPTX/AMDGPU enabled without having to also manually specify the CPU target.
I can try adding the original check and see how it goes

Reply:
> I guess it would make sense to automatically enable NVPTX/AMDGPU if the respective deps are specified in the EC or if cuda_compute_capabilities/amd_gfx_list are defined

For our systems at JSC, I added this small snippet to our "custom" LLVM EasyBlock and used your second idea to check if NVPTX should be built:

```python
cuda_cc_list = build_option('cuda_compute_capabilities') or self.cfg['cuda_compute_capabilities']
if cuda_cc_list:
    build_targets += ["NVPTX"]
    self.log.debug("Add NVPTX as additional build target architecture")
```

The code is not perfect, but is likely sufficient for our use case. LLVM does not require having CUDA / ROCm to build the respective NVPTX / AMDGPU backends as far as I am aware of, only when actually using them. 

> I am thinking the advantage of the automatic detection would be to be able to have NVPTX/AMDGPU enabled without having to also manually specify the CPU target.

Agreed, this was my thought as well. Especially if one has a "global" repository for multiple systems with different architectures, needing to specify the CPU target each time might get messy. Users are also still able to specify the targets manually, as the automatic detection is only done when `build_targets` is not defined.

Reply:
I've implemented the automatic detection based on having `cuda_compute_capabilities` or `amd_gfx_list` defined ([cac8fe9](https://github.com/easybuilders/easybuild-easyblocks/pull/3373/commits/cac8fe94162462a7b3860b5eaf2c38ed356f9995)).

For now this is being done only if no `build_targets` are being manually specified.
It is also true that if someone is specifying either of the 2 variables they might want to have the target enabled, but i feel giving priority should still be given to  `build_targets` (Maybe could add a check for printing a warning or outright raising an error if cuda_cc/amd_gfx are defined and the corresponding targets are not).

Reply:
I think your current approach is a good idea, only enabling the backends if the user has not defined which ones to build. Adding a warning (or at least a log entry) if he might've missed one might be a good idea.